tests/testthat/test-method-ops.R

```diff
@@ -1,60 +1,87 @@
-test_that("Ops generics dispatch to S7 methods", {
+test_that("Ops generics dispatch to S7 methods for S7 classes", {
+  local_methods(base_ops[["+"]])
+  foo1 <- new_class("foo1")
+  foo2 <- new_class("foo2")
+
+  method(`+`, list(foo1, foo1)) <- function(e1, e2) "foo1-foo1"
+  method(`+`, list(foo1, foo2)) <- function(e1, e2) "foo1-foo2"
+  method(`+`, list(foo2, foo1)) <- function(e1, e2) "foo2-foo1"
+  method(`+`, list(foo2, foo2)) <- function(e1, e2) "foo2-foo2"
+
+  expect_equal(foo1() + foo1(), "foo1-foo1")
+  expect_equal(foo1() + foo2(), "foo1-foo2")
+  expect_equal(foo2() + foo1(), "foo2-foo1")
+  expect_equal(foo2() + foo2(), "foo2-foo2")
+})
+
+test_that("Ops generics dispatch to S3 methods", {
+  skip_if(getRversion() < "4.3")
+  local_methods(base_ops[["+"]])
+
+  foo <- new_class("foo")
+  method(`+`, list(class_factor, foo)) <- function(e1, e2) "factor-foo"
+  method(`+`, list(foo, class_factor)) <- function(e1, e2) "foo-factor"
+
+  expect_equal(foo() + factor(), "foo-factor")
+  expect_equal(factor() + foo(), "factor-foo")
+
+  # Even if custom method exists
+  foo_S3 <- structure(list(), class = "foo_S3")
+  assign("+.foo_S3", function(e1, e2) stop("Failure!"), envir = globalenv())
+  on.exit(rm("+.foo_S3", envir = globalenv()))
```

Comment:
Does the `on.exit(add=FALSE)` default clear out the `defer()` call in the `local_methods()` call above?

Reply:
ooh yes